### PR TITLE
Give more control over echoing

### PR DESCRIPTION
**Pull request by hadley**
So you can now echo all input and output.

### CODE REVIEW

File: R/chat.R [hunk 1]
```diff
@@ -272,13 +278,19 @@ Chat <- R6::R6Class("Chat",
...
         if (any_text) {
           emit("\n")
           yield("\n")
         }
 
-        turn <- stream_turn(private$provider, result)
+        if (echo == "all") {
+          is_text <- map_lgl(turn@content, S7_inherits, content_text)
+          formatted <- map_chr(turn@content[!is_text], format)
+          cat_line(formatted, prefix = "< ")
```
Comment (jcheng5) [inline R/chat.R:292]: There are a number of places in this file where you're using `cat_line(..., prefix = "< ")`. Should these not all be word-wrapped? (`cat_word_wrap(prefix = "< ")(formatted)`)

File: R/chat.R [hunk 2]
```diff
@@ -66,19 +69,14 @@ Chat <- R6::R6Class("Chat",
...
       turn <- user_turn(...)
-      echo <- echo %||% private$echo
+      echo <- check_echo(echo %||% private$echo)
 
       # Returns a single turn (the final response from the assistant), even if
       # multiple rounds of back and forth happened.
-      coro::collect(private$chat_impl(turn, stream = echo, echo = echo))
+      coro::collect(private$chat_impl(turn, stream = echo != "none", echo = echo))
 
       text <- self$last_turn()@text
```
[the remark this answers is not in the document]
Comment (jcheng5) [reply R/chat.R:78]: FWIW we are mostly dumping Markdown at the console as well. At least GPT and Claude seem to generate it by default, and we're just catting it out.

File: R/chat.R (hunk 1 above)
Comment (jcheng5) [reply R/chat.R:292]: I went ahead and made `cat_line` do word wrapping. If there's some reason it shouldn't, feel free to undo that part.

File: R/chat.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (hadley) [reply R/chat.R:78]: @atheriel what should work for you in the short term is to ignore the output from `$chat()` and just retrieve the last turn with `$last_turn()`. Then you can do what ever you need to do with it. 

We'll need to introduce some new abstraction for retrieving different data types once we add structured data extraction, but I don't have any sense of what that will look like yet.

File: R/chat.R (hunk 1 above)
Comment (hadley) [reply R/chat.R:292]: Good call, thanks.

File: R/chat.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (atheriel) [reply R/chat.R:90]: Yeah, that makes sense. I'll dump a Markdown-formatted representation by default but Shiny apps could retrieve the structured data via `last_turn()`.

**Pull request merged**